### PR TITLE
As a user, I can save a game to a database

### PR DESCRIPTION
**Pull request by itsellej**
- The player's marks, player types and the squares as an array is saved to the database
- User can opt to resume the current game after saving it
- Fix - Board, ConsoleWriter & Grid - no longer implements Serializable

### CODE REVIEW

File: src/main/java/tictactoe/Controller.java [hunk 1]
```diff
@@ -2,27 +2,79 @@
...
-    public Controller(GameFactory gameFactory, Display display) {
+    public Controller(GameFactory gameFactory, Display display,
+                      GameSaver gameSaver, InputValidator inputValidator) {
         this.gameFactory = gameFactory;
         this.display = display;
+        this.gameSaver = gameSaver;
+        this.inputValidator = inputValidator;
     }
 
     public void playGame() {
```
Comment (Kotauror) [inline src/main/java/tictactoe/Controller.java]: I like how it's included in the existing logic - it's very easy to follow!

File: src/main/java/tictactoe/GameSaver.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    private void addEntryToDatabase
+            (ArrayList<String> gameState, String[] squares) {
+        var query = "INSERT INTO saved_games ("
+                + " game_name,"
+                + " current_player_mark,"
+                + " current_player_type,"
+                + " opponent_mark,"
+                + " opponent_type,"
+                + " squares ) VALUES ("
```
Comment (Kotauror) [inline src/main/java/tictactoe/GameSaver.java]: So it's squares at the end of the day :D I think it was the right decision.

File: src/main/java/tictactoe/GameSaver.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                + "?, ?, ?, ?, ?, ?)";
+
+        try {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            var squaresArray = connection.createArrayOf("text", squares);
+
+            preparedStatement.setString(1, gameState.get(0));
```
Comment (Kotauror) [inline src/main/java/tictactoe/GameSaver.java]: There is a lot of repetition going on there, did you think about using loop to rewrite it? And if not about loop, then about wrapping the `.setString` login in a method to make the code more descriptive?

File: src/main/java/tictactoe/GameSaver.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        } catch (SQLException ex) {
+            System.out.println(ex.getMessage());
+        }
+    }
+
+    public boolean isGameNameInDatabase(String gameName) {
+        var query = "SELECT * FROM saved_games WHERE game_name = ?";
+
+        try {
+            var retrievedGameName = "";
```
Comment (Kotauror) [inline src/main/java/tictactoe/GameSaver.java]: Did you think about wrapping everything that is inside of `try` into a private method?

File: src/main/java/tictactoe/InputValidator.java [hunk 5]
```diff
@@ -15,9 +15,14 @@ public InputValidator(BufferedReader bufferedReader, Display display) {
     public int validateMove(Grid grid, String playersMark, boolean isValid) {
         display.humanPlayerMessages(grid, playersMark, isValid);
         var input = getInput();
-        return (input.matches(ValidMove.DIGIT.move)) ? Integer.parseInt(input) : 0;
+        if (input.toLowerCase().matches(GameOptions.save.toString())) {
+            return -2;
```
Comment (Kotauror) [inline src/main/java/tictactoe/InputValidator.java]: why `-2`? It looks like a magic string to me.

File: src/test/java/tictactoe/ControllerTest.java [hunk 6]
```diff
@@ -1,14 +1,34 @@
...
 class ControllerTest {
 
+    @BeforeEach
+    void clearTable() {
+        var database = new Database();
+        var query = "DELETE FROM saved_games";
+
+        try {
+            var conn = database.connect();
+            PreparedStatement preparedStatement = conn.prepareStatement(query);
```
Comment (Kotauror) [inline src/test/java/tictactoe/ControllerTest.java]: I see this piece of logic being repeated - I'd suggest wrapping it into a method.

File: src/test/java/tictactoe/GameSaverTest.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class GameSaverTest {
+    Database database = new Database();
+    Connection connection = database.connect();
+    GameSaver gameSaver = new GameSaver(connection);
+
+    void clearTable() {
```
Comment (Kotauror) [inline src/test/java/tictactoe/GameSaverTest.java]: What do you think about passing table name as an argument to `clearTable`?

File: src/main/java/tictactoe/GameSaver.java (hunk 3 above)
Comment (itsellej) [reply src/main/java/tictactoe/GameSaver.java]: Thanks, Justyna. I'll look into this :)

File: src/main/java/tictactoe/GameSaver.java (hunk 4 above)
Comment (itsellej) [reply src/main/java/tictactoe/GameSaver.java]: Thanks, I'll look into this

File: src/test/java/tictactoe/GameSaverTest.java (hunk 7 above)
Comment (itsellej) [reply src/test/java/tictactoe/GameSaverTest.java]: I tried doing this and adding it to the query with string interpolation, however I read doing this is a SQL injection vulnerability (not sure if that's right or not). Do you have any suggestions for doing this? @kotauror

**Pull request merged**